osquery/carver/carver.cpp

```diff
@@ -0,0 +1,286 @@
+const std::string kCarveNamePrefix = "carve_";
+
+/// Database prefix used to directly access and manipulate our carver entries
+const std::string kCarverDBPrefix = "carving.";
+
+/*
+ * Carver block chunking size. This specifies how big of a block size to
+ * use when carving a file from disk.
+ */
+const size_t kBuffSize = 8192;
+
+DECLARE_string(tls_hostname);
+
+/// Session creation endpoint for forensic file carving
+CLI_FLAG(string,
+         carver_start_endpoint,
+         "",
+         "TLS/HTTPS init endpoint for forensic carver");
+
+/// Data aggregation endpoint for forensic file carving
+CLI_FLAG(
+    string,
+    carver_continue_endpoint,
+    "",
+    "TLS/HTTPS endpoint that receives carved content after session creation");
+
+/// Size of blocks used for POSTing data back to remote endpoints
+CLI_FLAG(int32,
+         carver_block_size,
+         8192,
+         "Size of blocks used for POSTing data back to remote endpoints");
+
+/// Helper function to update values related to a carve
+void updateCarveValue(const std::string& guid,
+                      const std::string& key,
+                      const std::string& value) {
+  std::string carve;
+  auto s = getDatabaseValue(kQueries, kCarverDBPrefix + guid, carve);
+  if (!s.ok()) {
+    VLOG(1) << "Unable to update status of carve " << guid;
+    return;
+  }
+
+  pt::ptree tree;
+  try {
+    std::stringstream ss(carve);
+    pt::read_json(ss, tree);
+  } catch (const pt::ptree_error& e) {
+    VLOG(1) << "Failed to parse carving entries: " << e.what();
+    return;
+  }
+
+  tree.put(key, value);
+
+  std::ostringstream os;
+  pt::write_json(os, tree, false);
+  setDatabaseValue(kQueries, kCarverDBPrefix + guid, os.str());
+}
+
+Carver::Carver(const std::set<std::string>& paths, const std::string& guid) {
+  for (const auto& p : paths) {
+    carvePaths_.insert(fs::path(p));
+  }
+
+  // Construct the uri we post our data back to:
+  startUri_ = TLSRequestHelper::makeURI(FLAGS_carver_start_endpoint);
+  contUri_ = TLSRequestHelper::makeURI(FLAGS_carver_continue_endpoint);
+
+  // Generate a unique identifier for this carve
+  carveGuid_ = guid;
+
+  // TODO: Adding in a manifest file of all carved files might be nice.
+  carveDir_ =
+      fs::temp_directory_path() / fs::path(kCarvePathPrefix + carveGuid_);
+  auto ret = fs::create_directory(carveDir_);
+  if (!ret) {
+    LOG(ERROR) << "Unable to create carve file store";
+  }
+
+  // Store the path to our archive for later exfiltration
+  archivePath_ = carveDir_ / fs::path(kCarveNamePrefix + carveGuid_ + ".tgz");
+
+  // Update the DB to reflect that the carve is pending.
+  updateCarveValue(carveGuid_, "status", "PENDING");
+};
+
+Carver::~Carver() {
+  fs::remove_all(carveDir_);
+}
+
+void Carver::start() {
+  for (const auto& p : carvePaths_) {
+    if (!fs::exists(p)) {
+      LOG(WARNING) << "File does not exist on disk: " << p;
+    } else {
+      Status s = carve(p);
+      if (!s.ok()) {
```

Comment:
Similar to the above. If a single file fails to be carved we might still get other files.